src/TaskManager/TaskManager.cs

```diff
@@ -198,15 +202,18 @@ private async Task HandleTaskCallback(JsonMessage<TaskCallbackEvent> message)
                 {
                     var metadata = await metadataRepository.RetrieveMetadata().ConfigureAwait(false);
 
-                    foreach (var item in metadata)
-                        updateMessage.Body.Metadata.Add(item.Key, item.Value);
+                    if (metadata is not null)
+                    {
+                        foreach (var item in metadata)
+                            updateMessage.Body.Metadata.Add(item.Key, item.Value);
+                    }
                 }
                 catch (Exception ex)
                 {
-                    updateMessage.Body.Status = TaskExecutionStatus.Failed;
-                    updateMessage.Body.Reason = FailureReason.PluginError;
-
                     _logger.MetadataRetrievalFailed(ex);
+                    await HandleMessageException(message, message.Body.WorkflowInstanceId, message.Body.TaskId, message.Body.ExecutionId, ex.Message, false).ConfigureAwait(false);
+
+                    return;
                 }
             }
```

Comment:
this whole section feels like its adding a lot of complexity to handle task callback which is like 100 lines long, could it be moved to a separate function from seems like its all to do with metadata from that point?

line 184 
`            if (!string.IsNullOrWhiteSpace(metadataAssembly))`

Reply:
Pulled what i could out into a separate method, looks cleaner now